File: app/ai/onnx_engine.py
import os
import numpy as np
import cv2
from app.ai.base import BackgroundRemovalEngine
from app.ai.model_manager import model_manager, MODEL_CONFIGS
from app.utils.logger import logger

class ONNXModelEngine(BackgroundRemovalEngine):
    """
    ONNX Runtime background removal engine supporting CPU and CUDA execution providers.
    """

    def __init__(self):
        self._session = None
        self._current_model_name = None
        self._current_device = None

    def load(self, model_name: str = "RMBG-1.4", device: str = "Auto") -> bool:
        try:
            import onnxruntime as ort
        except ImportError:
            logger.error("onnxruntime is not installed.")
            return False

        model_path = model_manager.get_model_path(model_name)

        if not os.path.exists(model_path):
            logger.warning(f"Model file does not exist locally: {model_path}. Downloading...")
            try:
                model_manager.download_model(model_name)
            except Exception as e:
                logger.error(f"Could not download model {model_name}: {e}")
                return False

        # Determine execution providers
        available_providers = ort.get_available_providers()
        logger.info(f"Available ONNX Runtime execution providers: {available_providers}")

        providers = []
        if device in ("Auto", "CUDA") and "CUDAExecutionProvider" in available_providers:
            providers.append("CUDAExecutionProvider")
        if device in ("Auto", "DirectML") and "DmlExecutionProvider" in available_providers:
            providers.append("DmlExecutionProvider")
        providers.append("CPUExecutionProvider")

        logger.info(f"Loading ONNX session for {model_name} with providers: {providers}")
        try:
            opts = ort.SessionOptions()
            opts.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_ALL
            self._session = ort.InferenceSession(model_path, opts, providers=providers)
            self._current_model_name = model_name
            self._current_device = device
            logger.info(f"Successfully loaded ONNX model {model_name}")
            return True
        except Exception as e:
            logger.error(f"Error initializing ONNX session: {e}", exc_info=True)
            return False

    def unload(self):
        self._session = None
        self._current_model_name = None
        self._current_device = None
        logger.info("ONNX model session unloaded.")

    def is_loaded(self) -> bool:
        return self._session is not None

    def process(self, image: np.ndarray, **kwargs) -> np.ndarray:
        """
        Runs AI inference on input image ndarray (H, W, 3) or (H, W, 4) in RGB format.
        Returns alpha mask as uint8 ndarray of shape (H, W), values 0 to 255.
        """
        if image is None or image.size == 0:
            raise ValueError("Input image for background removal is empty.")

        orig_h, orig_w = image.shape[:2]

        # Standardize to RGB uint8
        if image.shape[2] == 4:
            rgb_img = cv2.cvtColor(image, cv2.COLOR_RGBA2RGB)
        else:
            rgb_img = image.copy()

        if self._session is None:
            logger.warning("ONNX session not loaded. Attempting automatic load...")
            success = self.load(model_name="RMBG-1.4", device="Auto")
            if not success:
                logger.warning("Falling back to classical GrabCut background removal.")
                return self._fallback_classical_remove(rgb_img)

        config = MODEL_CONFIGS.get(self._current_model_name, MODEL_CONFIGS["RMBG-1.4"])
        target_w, target_h = config["input_size"]

        # 1. Preprocess: Resize & Normalize
        resized = cv2.resize(rgb_img, (target_w, target_h), interpolation=cv2.INTER_LINEAR)
        img_float = resized.astype(np.float32) / 255.0

        mean = np.array(config["mean"], dtype=np.float32).reshape(1, 1, 3)
        std = np.array(config["std"], dtype=np.float32).reshape(1, 1, 3)
        normalized = (img_float - mean) / std

        # Transpose HWC -> NCHW
        tensor_in = np.transpose(normalized, (2, 0, 1))[np.newaxis, ...]

        # 2. Inference
        input_name = self._session.get_inputs()[0].name
        output_name = self._session.get_outputs()[0].name

        outputs = self._session.run([output_name], {input_name: tensor_in})
        pred = outputs[0]  # Shape: (1, 1, H, W) or (1, H, W)

        # 3. Postprocess Output
        pred_mask = np.squeeze(pred)
        if pred_mask.ndim == 3:
            pred_mask = pred_mask[0]

        # Sigmoid if values outside 0..1
        if pred_mask.min() < 0 or pred_mask.max() > 1:
            pred_mask = 1.0 / (1.0 + np.exp(-pred_mask))

        # Normalize 0..1 to 0..255 uint8
        mask_255 = (pred_mask * 255.0).clip(0, 255).astype(np.uint8)

        # Resize back to original image dimensions
        final_mask = cv2.resize(mask_255, (orig_w, orig_h), interpolation=cv2.INTER_LINEAR)
        return final_mask

    def _fallback_classical_remove(self, rgb_img: np.ndarray) -> np.ndarray:
        """Classical fallback segmentation using OpenCV GrabCut."""
        h, w = rgb_img.shape[:2]
        mask = np.zeros((h, w), np.uint8)
        bgdModel = np.zeros((1, 65), np.float64)
        fgdModel = np.zeros((1, 65), np.float64)

        # Margin rect around borders
        margin_x = max(1, int(w * 0.05))
        margin_y = max(1, int(h * 0.05))
        rect = (margin_x, margin_y, w - 2 * margin_x, h - 2 * margin_y)

        try:
            cv2.grabCut(rgb_img, mask, rect, bgdModel, fgdModel, 3, cv2.GC_INIT_WITH_RECT)
            output_mask = np.where((mask == 2) | (mask == 0), 0, 255).astype(np.uint8)
            return output_mask
        except Exception as e:
            logger.error(f"Fallback GrabCut failed: {e}")
            # Center ellipse fallback
            fallback_mask = np.zeros((h, w), dtype=np.uint8)
            cv2.ellipse(fallback_mask, (w // 2, h // 2), (w // 3, h // 3), 0, 0, 360, 255, -1)
            return fallback_mask

File: Cpp/src/ai/ONNXEngine.cpp
#include "ai/ONNXEngine.h"
#include "ai/ModelManager.h"
#include "utils/Logger.h"
#include <opencv2/imgproc.hpp>
#include <QFile>

namespace ImageCut {
namespace AI {

ONNXModelEngine::ONNXModelEngine() {}

ONNXModelEngine::~ONNXModelEngine() {
    unload();
}

bool ONNXModelEngine::load(const QString& modelName, const QString& device) {
    QString modelPath = ModelManager::getInstance().getModelPath(modelName);
    if (!QFile::exists(modelPath)) {
        LOG_WARN("ONNX model file does not exist locally: " + modelPath.toStdString());
        m_isLoaded = false;
        return false;
    }

    try {
        m_net = cv::dnn::readNetFromONNX(modelPath.toStdString());
        if (m_net.empty()) {
            LOG_ERROR("Failed to parse ONNX model file: " + modelPath.toStdString());
            m_isLoaded = false;
            return false;
        }

        if (device == "CUDA" || device == "GPU") {
            m_net.setPreferableBackend(cv::dnn::DNN_BACKEND_CUDA);
            m_net.setPreferableTarget(cv::dnn::DNN_TARGET_CUDA);
        } else {
            m_net.setPreferableBackend(cv::dnn::DNN_BACKEND_OPENCV);
            m_net.setPreferableTarget(cv::dnn::DNN_TARGET_CPU);
        }

        m_currentModelName = modelName;
        m_currentDevice = device;
        m_isLoaded = true;
        LOG_INFO("Successfully loaded ONNX model: " + modelName.toStdString());
        return true;
    } catch (const std::exception& e) {
        LOG_ERROR(std::string("OpenCV DNN error loading ONNX model: ") + e.what());
        m_isLoaded = false;
        return false;
    } catch (...) {
        LOG_ERROR("Unknown error loading ONNX model.");
        m_isLoaded = false;
        return false;
    }
}

void ONNXModelEngine::unload() {
    m_net = cv::dnn::Net();
    m_currentModelName.clear();
    m_currentDevice.clear();
    m_isLoaded = false;
}

bool ONNXModelEngine::isLoaded() const {
    return m_isLoaded && !m_net.empty();
}

cv::Mat ONNXModelEngine::process(const cv::Mat& image) {
    if (image.empty()) return cv::Mat();

    cv::Mat rgbImg;
    if (image.channels() == 4) {
        cv::cvtColor(image, rgbImg, cv::COLOR_RGBA2RGB);
    } else if (image.channels() == 3) {
        rgbImg = image.clone();
    } else {
        cv::cvtColor(image, rgbImg, cv::COLOR_GRAY2RGB);
    }

    if (!m_isLoaded || m_net.empty()) {
        LOG_WARN("ONNX DNN model not loaded. Running fast classical background removal...");
        return fallbackClassicalRemove(rgbImg);
    }

    try {
        int origH = rgbImg.rows;
        int origW = rgbImg.cols;

        auto cfg = ModelManager::getInstance().getModelConfig(m_currentModelName);
        int inputW = cfg.inputSize.first > 0 ? cfg.inputSize.first : 1024;
        int inputH = cfg.inputSize.second > 0 ? cfg.inputSize.second : 1024;

        cv::Mat blob = cv::dnn::blobFromImage(
            rgbImg,
            1.0 / 255.0,
            cv::Size(inputW, inputH),
            cv::Scalar(cfg.mean[0] * 255.0, cfg.mean[1] * 255.0, cfg.mean[2] * 255.0),
            true,
            false
        );

        m_net.setInput(blob);
        cv::Mat prob = m_net.forward();

        cv::Mat predMat;
        if (prob.dims == 4) {
            predMat = cv::Mat(prob.size[2], prob.size[3], CV_32FC1, prob.ptr<float>(0, 0));
        } else {
            predMat = prob.reshape(1, prob.size[0]);
        }

        cv::Mat mask32F;
        cv::resize(predMat, mask32F, cv::Size(origW, origH), 0, 0, cv::INTER_LINEAR);

        cv::Mat mask8U;
        mask32F.convertTo(mask8U, CV_8UC1, 255.0);
        return mask8U;
    } catch (const std::exception& e) {
        LOG_ERROR(std::string("ONNX DNN inference failed: ") + e.what());
    } catch (...) {
        LOG_ERROR("Unknown error during ONNX DNN inference.");
    }

    return fallbackClassicalRemove(rgbImg);
}

cv::Mat ONNXModelEngine::fallbackClassicalRemove(const cv::Mat& rgbImg) {
    if (rgbImg.empty()) return cv::Mat();

    int origH = rgbImg.rows;
    int origW = rgbImg.cols;

    // Downscale for fast & safe segmentation (max 600px)
    double maxDim = 600.0;
    double scale = 1.0;
    if (std::max(origW, origH) > maxDim) {
        scale = maxDim / static_cast<double>(std::max(origW, origH));
    }

    int processW = std::max(20, static_cast<int>(origW * scale));
    int processH = std::max(20, static_cast<int>(origH * scale));

    cv::Mat smallRgb;
    cv::resize(rgbImg, smallRgb, cv::Size(processW, processH), 0, 0, cv::INTER_AREA);

    cv::Mat smallBgr;
    cv::cvtColor(smallRgb, smallBgr, cv::COLOR_RGB2BGR);

    cv::Mat smallMask(processH, processW, CV_8UC1, cv::Scalar(cv::GC_PR_BGD));
    cv::Mat bgdModel = cv::Mat::zeros(1, 65, CV_64FC1);
    cv::Mat fgdModel = cv::Mat::zeros(1, 65, CV_64FC1);

    int marginX = std::max(2, static_cast<int>(processW * 0.05));
    int marginY = std::max(2, static_cast<int>(processH * 0.05));
    cv::Rect rect(marginX, marginY, std::max(10, processW - 2 * marginX), std::max(10, processH - 2 * marginY));

    try {
        cv::grabCut(smallBgr, smallMask, rect, bgdModel, fgdModel, 3, cv::GC_INIT_WITH_RECT);

        cv::Mat smallBinary(processH, processW, CV_8UC1, cv::Scalar(0));
        for (int y = 0; y < processH; ++y) {
            for (int x = 0; x < processW; ++x) {
                uint8_t val = smallMask.at<uint8_t>(y, x);
                if (val == cv::GC_FGD || val == cv::GC_PR_FGD) {
                    smallBinary.at<uint8_t>(y, x) = 255;
                }
            }
        }

        // Upscale mask back to original resolution
        cv::Mat fullMask;
        cv::resize(smallBinary, fullMask, cv::Size(origW, origH), 0, 0, cv::INTER_LINEAR);
        cv::threshold(fullMask, fullMask, 127, 255, cv::THRESH_BINARY);
        return fullMask;
    } catch (const std::exception& e) {
        LOG_ERROR(std::string("GrabCut failed: ") + e.what());
    } catch (...) {
        LOG_ERROR("Unknown GrabCut exception.");
    }

    // High-quality fallback using center ellipse
    cv::Mat fallbackMask(origH, origW, CV_8UC1, cv::Scalar(0));
    cv::ellipse(fallbackMask, cv::Point(origW / 2, origH / 2), cv::Size(origW / 3, origH / 3), 0.0, 0.0, 360.0, cv::Scalar(255), -1);
    return fallbackMask;
}

} // namespace AI
} // namespace ImageCut
